src/asymmetric_crypto/rsa/private_key.rs

```diff
@@ -0,0 +1,151 @@
+            }
+            RsaKeyWrappingAlgorithm::Aes256Sha1 => {
+                ckm_rsa_aes_key_unwrap::<sha1::Sha1>(self, encrypted_key_material)?
+            }
+            RsaKeyWrappingAlgorithm::Aes256Sha3 => {
+                ckm_rsa_aes_key_unwrap::<sha3::Sha3_256>(self, encrypted_key_material)?
+            }
+        })
+    }
+}
+
+impl PrivateKey for RsaPrivateKey {
+    type PublicKey = RsaPublicKey;
+
+    fn public_key(&self) -> Self::PublicKey {
+        self.0.to_public_key().into()
+    }
+}
+
+/// Facades
+impl RsaPrivateKey {
+    /// Get the public key
+    ///
+    /// This is a facade for [`PrivateKey::public_key`]
+    pub fn public_key(&self) -> <RsaPrivateKey as crate::asymmetric_crypto::PrivateKey>::PublicKey {
+        PrivateKey::public_key(self)
+    }
+}
+
+impl ZeroizeOnDrop for RsaPrivateKey {}
+
+/// Implementation of PKCS#1 RSA OAEP (CKM_RSA_PKCS_OAEP)
+/// [https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/os/pkcs11-curr-v3.0-os.html#_Toc30061137]
+fn ckm_rsa_pkcs_oaep<H: 'static + Digest + DynDigest + Send + Sync>(
+    rsa_private_key: &RsaPrivateKey,
+    encrypted_key_material: &[u8],
+) -> Result<Zeroizing<Vec<u8>>, CryptoCoreError> {
+    let padding = rsa::Oaep::new::<H>();
+    Ok(Zeroizing::from(
+        rsa_private_key.0.decrypt(padding, encrypted_key_material)?,
+    ))
+}
+
+/// Implementation of PKCS#11 RSA AES KEY WRAP (CKM_RSA_AES_KEY_WRAP)
+/// [https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/os/pkcs11-curr-v3.0-os.html#_Toc30061152]
+///
+/// Also check [https://cloud.google.com/kms/docs/key-wrapping?hl=fr]
+fn ckm_rsa_aes_key_unwrap<H: 'static + Digest + DynDigest + Send + Sync>(
+    rsa_private_key: &RsaPrivateKey,
+    ciphertext: &[u8],
+) -> Result<Zeroizing<Vec<u8>>, CryptoCoreError> {
+    // Splits the input into two parts. The first is the wrapped AES key, and the
+    // second is the wrapped target key. The length of the first part is equal to
+    // the length of the unwrapping RSA key.
+    let encapsulation_bytes_len = ((rsa_private_key.key_length() as i32) / 8) as usize;
+    // Un-wraps the temporary AES key from the first part with the private RSA key
+    // using CKM_RSA_PKCS_OAEP with parameters of OAEPParams.
+    let aes_key = ckm_rsa_pkcs_oaep::<H>(
+        rsa_private_key,
+        ciphertext[0..encapsulation_bytes_len].as_ref(),
+    )?;
+    // Un-wraps the target key from the second part with the temporary AES key using
+    // CKM_AES_KEY_WRAP_KWP ([AES KEYWRAP] section 6.3). Zeroizes the temporary
+    // AES key. Returns the handle to the newly unwrapped target key.
+    Ok(Zeroizing::from(key_unwrap(
+        &ciphertext[encapsulation_bytes_len..],
+        &aes_key,
+    )?))
+}
+
+impl pkcs8::EncodePrivateKey for RsaPrivateKey {
+    fn to_pkcs8_der(&self) -> pkcs8::Result<SecretDocument> {
+        self.0.to_pkcs8_der()
+    }
+
+    fn to_pkcs8_encrypted_der(
+        &self,
+        rng: impl rand_core::CryptoRng + RngCore,
```

Comment:
The trait `CryptoRngCore` is made to avoid that. Plus, you shouldn't take ownership of the `rng`.
```suggestion
        rng: &mut impl CryptoRngCore,
```